### PR TITLE
WT-10541 Implement exception handling for Azure Connection class

### PR DESCRIPTION
**Pull request by lting0804**
- Implement exception handling for Azure Connection -> maps the HTTP error code if one is thrown to an OS error. If not the exception is caught and the reason is just printed out.
- Created free function to deal with exceptions thrown by Azure

### CODE REVIEW

File: ext/storage_sources/azure_store/azure_connection.cpp [hunk 1]
```diff
@@ -60,6 +60,15 @@ azure_connection::list_objects(
...
+    try {
+        _azure_client.ListBlobs(blob_parameters);
+    } catch (const Azure::Core::RequestFailedException &e) {
+        return http_to_errno(e);
+    } catch (const std::exception &e) {
+        std::cerr << e.what() << std::endl;
+        return -1;
+    }
+
     auto list_blobs_response = _azure_client.ListBlobs(blob_parameters);
```
Comment (mariammojid) [inline ext/storage_sources/azure_store/azure_connection.cpp:72]: could you include the `list_blobs_response` in the `try` block? To reduce code duplication.

**Review by mariammojid — COMMENTED** [review]
Looks good, just a minor suggestion

File: ext/storage_sources/azure_store/azure_connection.h [hunk 2]
```diff
@@ -33,7 +33,15 @@
...
-#include <iostream>
+
+// Mapping between HTTP response codes and corresponding errno values to be used by the Azure
+// connection methods to return errno values expected by the filesystem interface.
+static const std::map<Azure::Core::Http::HttpStatusCode, int32_t> to_errno = {
+  {Azure::Core::Http::HttpStatusCode::NotFound, ENOENT},
+  {Azure::Core::Http::HttpStatusCode::Forbidden, EACCES},
+  {Azure::Core::Http::HttpStatusCode::Conflict, EBUSY},
+  {Azure::Core::Http::HttpStatusCode::BadRequest, EINVAL},
+  {Azure::Core::Http::HttpStatusCode::InternalServerError, EAGAIN}};
```
Comment (korteland) [inline ext/storage_sources/azure_store/azure_connection.h:44]: I really like this approach

**Review by korteland — APPROVED** [review]
LGTM - nice work!

File: ext/storage_sources/azure_store/azure_connection.cpp [hunk 3]
```diff
@@ -73,9 +82,16 @@ int
 azure_connection::put_object(const std::string &object_key, const std::string &file_path) const
 {
     auto blob_client = _azure_client.GetBlockBlobClient(_object_prefix + object_key);
-    // UploadFrom will always return a UploadBlockBlobFromResult describing the state of the updated
-    // block blob so there's no need to check for errors.
-    blob_client.UploadFrom(file_path);
+    // UploadFrom will return a UploadBlockBlobFromResult describing the state of the updated
```
Comment (tammybailey) [inline ext/storage_sources/azure_store/azure_connection.cpp:85]: Can we update this comment slightly?
```
// UploadFrom returns a UploadBlockBlobFromResult object describing the state of the updated
// block blob on success and throws an exception on failure.
```

File: ext/storage_sources/azure_store/azure_connection.cpp [hunk 4]
```diff
@@ -86,14 +102,16 @@ azure_connection::delete_object(const std::string &object_key) const
     std::string obj = _object_prefix + object_key;
 
     auto object_client = _azure_client.GetBlobClient(obj);
-    auto delete_blob_response = object_client.DeleteIfExists();
 
-    // Returns false if obj doesn't exist.
-    if (!delete_blob_response.Value.Deleted) {
-        std::cerr << obj + " : No such object file exists in the bucket." << std::endl;
+    // Delete will delete the blob if it exists or an exception.
```
Comment (tammybailey) [inline ext/storage_sources/azure_store/azure_connection.cpp:106]: Similar to above:
```
// Delete will delete the blob if it exists or throw an exception on failure.
```

File: ext/storage_sources/azure_store/azure_connection.cpp [hunk 5]
```diff
@@ -103,6 +121,17 @@ azure_connection::read_object(
   const std::string &object_key, int64_t offset, size_t len, void *buf) const
 {
     auto blob_client = _azure_client.GetBlockBlobClient(_object_prefix + object_key);
+
+    // GetProperties will return BlobProperties describing the blob or an exception.
```
Comment (tammybailey) [inline ext/storage_sources/azure_store/azure_connection.cpp:125]: We could add some detail to this comment to be more useful for the reader:
```
// GetProperties returns a BlobProperties object containing the blob size on success 
// and throws an exception on failure.
```

File: ext/storage_sources/azure_store/azure_connection.cpp [hunk 6]
```diff
@@ -178,3 +207,14 @@ azure_connection::bucket_exists(bool &exists) const
     }
     return 0;
 }
+
+// Maps the HTTP code returned by Azure SDK and returns the associated errno code if it is in
```
Comment (tammybailey) [inline ext/storage_sources/azure_store/azure_connection.cpp:211]: We can break up this run-on sentence into two:
```
// Return a system error code corresponding to the HTTP code returned by
// the Azure SDK. If no such mapping is defined, return -1.
```

**Review by tammybailey — APPROVED** [review]
Code looks good but comments could be improved. I've left some suggestions!

File: ext/storage_sources/azure_store/azure_connection.cpp (hunk 3 above)
Comment (lting0804) [reply ext/storage_sources/azure_store/azure_connection.cpp]: resolved here https://github.com/wiredtiger/wiredtiger/pull/8750/commits/e42f214727eda92b86bcea0d45fc6c00676173e8

File: ext/storage_sources/azure_store/azure_connection.cpp (hunk 4 above)
Comment (lting0804) [reply ext/storage_sources/azure_store/azure_connection.cpp]: resolved here https://github.com/wiredtiger/wiredtiger/pull/8750/commits/e42f214727eda92b86bcea0d45fc6c00676173e8

File: ext/storage_sources/azure_store/azure_connection.cpp (hunk 5 above)
Comment (lting0804) [reply ext/storage_sources/azure_store/azure_connection.cpp]: resolved here https://github.com/wiredtiger/wiredtiger/pull/8750/commits/e42f214727eda92b86bcea0d45fc6c00676173e8

File: ext/storage_sources/azure_store/azure_connection.cpp (hunk 6 above)
Comment (lting0804) [reply ext/storage_sources/azure_store/azure_connection.cpp]: resolved here https://github.com/wiredtiger/wiredtiger/pull/8750/commits/e42f214727eda92b86bcea0d45fc6c00676173e8

File: ext/storage_sources/azure_store/azure_connection.cpp (hunk 3 above)
Comment (lting0804) [reply ext/storage_sources/azure_store/azure_connection.cpp]: Thanks for the suggestions @tammybailey I've taken them onboard

**Comment by lting0804** [comment]
evergreen merge

**Pull request merged**